### PR TITLE
Develop ugwp improvements

### LINKED ISSUE #20: Improve stability of unified drag suite and add diagnostics
## Description
The unified gravity wave physics (UGWP) package has experienced numerical instability during FV3GFS Prototype 8 testing at C384 horizontal resolution.  (Model crashes are documented [here](https://docs.google.com/document/d/13TiaC6D-h0wdkiDck-vDJze-G67KFbHHgSS2kZzsCHo/edit) and [here.)](https://docs.google.com/presentation/d/1kkRHx4AVKl4Nn7yTiJoY3fyrkGlxpcRhhPp5G5jOrzU/edit#slide=id.g136848aa835_0_50)  The source of the instabilities has been traced to the mesoscale gravity wave drag (GWD) and turbulent orographic form drag (TOFD) parameterizations contained in the CCPP module drag_suite.F90.  The instability can be alleviated by taking a smaller physics time step, however, this is not a solution as it would not be amenable to operational forecasting.

## Solution
The solution is to limit the tendencies calculated by the mesoscale GWD scheme in drag_suite.F90 in a similar manner as is done in the GFSv16 scheme in gwdps.f.  The instability due to the TOFD parameterization was due to an earlier modification of the scheme that removed a limit on the subgrid terrain variability and caused excessively large tendencies.  The limits have been re-introduced to help alleviate the instability problem.
Another change to the code was the addition of optional diagnostic outputs, controlled by the namelist option "ldiag_ugwp", which outputs the drag contributions from each of the drag components of the UGWP:  mesoscale GWD, low-level blocking, small-scale GWD, TOFD and non-stationary GWD.

### PR DESCRIPTION
**Pull request by mdtoyNOAA**
## Description

This pull request has three main components:
1)  To fix some issues with the orographic drag schemes of the "_drag_suite_" CCPP module which have caused model crashes to occur in GFSv17 p8 testing.  These issues are described in [ufs-community/ccpp-physics/Issue #20](https://github.com/ufs-community/ccpp-physics/issues/20).
2)  To add optional diagnostic outputs for each of the components of the drag suite for tuning purposes.  This is also described in [ufs-community/ccpp-physics/Issue #20](https://github.com/ufs-community/ccpp-physics/issues/20).  The optional diagnostic outputs can be selected at run time by the already-existing _ldiag_ugwp_ namelist flag.
3)  As part of the 1st component above, at the suggestion of Philip Pegion, we have added the capability to update the u and v winds in each model column based on the tendencies calculated by the mesoscale gravity wave drag and blocking schemes, and using these for the calculation of the tendencies by the small-scale gravity wave drag and turbulent orographic form drag schemes.  This sequential updating of the wind profile has been shown to improve the numerical stability.  The option to update the winds in this manner is controlled by a new logic flag _ugwp_seq_update_, which for now, is not a namelist option, but is hardwired by default to ._false_. in _GFS_typedefs.F90_.  The user can change this to ._true_. at compile time.

Is a change of answers expected from this PR?  Yes.

### Issue(s) addressed

- fixes [ufs-community/ccpp-physics/Issue #20](https://github.com/ufs-community/ccpp-physics/issues/20)

## Testing

How were these changes tested?  RT (new baseline, then tested against new baseline).  ORT.
What compilers / HPCs was it tested with?  Intel.  Hera.

Have the ufs-weather-model regression test been run? Yes.
On what platform?  Hera/Intel.
- Will the code updates change regression test baseline? Yes.  Changes to CCPP unified_ugwp drag suite. Please show the baseline directory below.
/scratch1/BMC/wrfruc/mtoy/git_local/ufs-weather-model/REGRESSION_TEST_INTEL.2022_11_08

## Dependencies

If testing this branch requires non-default branches in other repositories, list them.
https://github.com/mdtoyNOAA/ufs-weather-model/tree/develop_ugwp_improvements
https://github.com/mdtoyNOAA/ccpp-physics/tree/ufs/dev_ugwp_improvements

CCPP-physics pull request dependency: [ccpp-physics#20](https://github.com/ufs-community/ccpp-physics/pull/22)

### CODE REVIEW

File: ccpp/data/GFS_typedefs.meta [hunk 1]
```diff
@@ -6323,6 +6323,12 @@
   units = flag
   dimensions = ()
   type = logical
+[ugwp_seq_update]
+  standard_name = ugwp_sequential_update_flag
```
Comment (grantfirl) [inline ccpp/data/GFS_typedefs.meta]: Could you change this new standard name to "do_ugwp_sequential_update" (here and in ccpp-physics) please?

File: ccpp/data/GFS_typedefs.meta [hunk 2]
```diff
@@ -6354,8 +6360,8 @@
   dimensions = ()
   type = logical
 [do_gsl_drag_ls_bl]
-  standard_name = flag_for_gsl_drag_suite_large_scale_orographic_and_blocking_drag
-  long_name = flag to activate GSL drag suite - large-scale GWD and blocking
+  standard_name = flag_for_gsl_drag_suite_mesoscale_orographic_and_blocking_drag
```
Comment (grantfirl) [inline ccpp/data/GFS_typedefs.meta]: Similarly, could you change this to "do_gsl_drag_suite_mesoscale_orographic_and_blocking_drag". We're trying to switch the names of these kind of flags to be more descriptive to match other CCPP models.

File: ccpp/data/GFS_typedefs.meta (hunk 1 above)
Comment (mdtoyNOAA) [reply ccpp/data/GFS_typedefs.meta]: Updated code and pushed to PR.

File: ccpp/data/GFS_typedefs.meta (hunk 2 above)
Comment (mdtoyNOAA) [reply ccpp/data/GFS_typedefs.meta]: Updated code and pushed to PR.

**Review by DusanJovic-NOAA — APPROVED** [review]

**Comment by ChunxiZhang-NOAA** [comment]
@mdtoyNOAA  Do you mind cleaning up some variables related to UGWP in GFS_typedefs.F90. Most unused variables are in this block:
if (Model%ldiag_ugwp) then
      ........
      dv3dt_mtb
      ......
endif

**Comment by mdtoyNOAA** [comment]
> @mdtoyNOAA Do you mind cleaning up some variables related to UGWP in GFS_typedefs.F90. Most unused variables are in this block:
> if (Model%ldiag_ugwp) then
> ........
> dv3dt_mtb
> ......
> endif

I have cleaned up several variables and am running RT's to make sure nothing is broken.
The unused variables I removed are:
dv3dt_ogw, dt3dt_ogw, dv3dt_mtb, dt3dt_mtb, dv3dt_tms, dt3dt_tms, dt3dt_ngw, du3dt_cgw, dv3dt_cgw, dt3dt_moist, du3dt_moist and dv3dt_moist

**Comment by ChunxiZhang-NOAA** [comment]
That's great, thank you Mike.

**Comment by ChunxiZhang-NOAA** [comment]
@mdtoyNOAA  @jkbk2004  ccpp framework didn't point to the correct hash. This PR points to the latest hash, but the upstream/develop branch uses an old version. But I think it is ok to update to the latest hash.

**Comment by jkbk2004** [comment]
@ChunxiZhang-NOAA nice catch! @DusanJovic-NOAA @grantfirl updating ccpp framework hash can be a right way to go, isn't it?

**Comment by DusanJovic-NOAA** [comment]
> @ChunxiZhang-NOAA nice catch! @DusanJovic-NOAA @grantfirl updating ccpp framework hash can be a right way to go, isn't it?

I guess. I'm not familiar with changes in ccpp-framework.

**Comment by mdtoyNOAA** [comment]
> > @ChunxiZhang-NOAA nice catch! @DusanJovic-NOAA @grantfirl updating ccpp framework hash can be a right way to go, isn't it?
> 
> I guess. I'm not familiar with changes in ccpp-framework.

The latest updates to the "main" branch of "NCAR/ccpp-framework" are:

commit d4a00031381afa7cd90589040a5c7e8f53991801 (HEAD -> main, origin/main, origin/HEAD)
Merge: 4daaa24 acd3b32
Author: Michael Kavulich <kavulich@ucar.edu>
Date:   Tue Oct 11 15:35:48 2022 -0600

    Merge pull request #455 from mkavulich/limit_cap_names
    
    Limit suite names to 37 characters or less

commit acd3b327ab449a98deb8916254573f537a5ce8d2
Author: Michael Kavulich, Jr <kavulich@ucar.edu>
Date:   Tue Sep 27 11:38:51 2022 -0600

    Limit suite names to 37 characters or less


Looks like minor changes that shouldn't change the results.

**Review by ChunxiZhang-NOAA — APPROVED** [review]
Looks good. Note that this PR will change the ccpp framework hash to d4a0003.

**Comment by jkbk2004** [comment]
@mdtoyNOAA ccpp pr was merged. can you update submodule pointer?

**Comment by jkbk2004** [comment]
@mdtoyNOAA revert branch in gitmodules as well.

**Comment by mdtoyNOAA** [comment]
Done.

**Pull request merged**